File: mns-reactor/src/test/java/cz/fb/manaus/reactor/rounding/MatchbookRoundingPluginTest.java
package cz.fb.manaus.reactor.rounding;

import cz.fb.manaus.core.provider.ExchangeProvider;
import cz.fb.manaus.core.test.AbstractLocalTestCase;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.context.ActiveProfiles;

import static cz.fb.manaus.spring.ManausProfiles.TEST_PROFILE;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

@ActiveProfiles(value = {"matchbook", TEST_PROFILE}, inheritProfiles = false)
public class MatchbookRoundingPluginTest extends AbstractLocalTestCase {

    @Autowired
    private MatchbookRoundingPlugin plugin;
    @Autowired
    private ExchangeProvider provider;


    @Test
    public void testShift() {
        assertEquals(2.02, plugin.shift(2, 1).getAsDouble(), 0.0001d);
    }

    @Test
    public void testInc1x() {
        assertEquals(1.227, plugin.shift(1.222d, 1).getAsDouble(), 0.001d);
        assertEquals(1.128, plugin.shift(1.125d, 1).getAsDouble(), 0.001d);
        assertEquals(1.131, plugin.shift(1.128d, 1).getAsDouble(), 0.001d);
        assertEquals(1.13, plugin.shift(1.125d, 2).getAsDouble(), 0.001d);
        assertEquals(2.01d, plugin.shift(1.99d, 1).getAsDouble(), 0.001d);
    }

    @Test
    public void testDec2x() {
        assertEquals(1.98d, plugin.shift(2d, -1).getAsDouble(), 0.0001d);
        assertEquals(2.764d, plugin.shift(2.8d, -1).getAsDouble(), 0.0001d);
    }

    @Test
    public void testDec3x() {
        assertEquals(2.96d, plugin.shift(3d, -1).getAsDouble(), 0.0001d);
    }

    @Test
    public void testInc3x() {
        assertEquals(3.55d, plugin.shift(3.5d, 1).getAsDouble(), 0.0001d);
    }

    @Test
    public void testInc4x() {
        assertEquals(5.08d, plugin.shift(5d, 1).getAsDouble(), 0.0001d);
    }

    @Test
    public void testRoundBigTest() {
        var previous = -1d;
        for (var price = provider.getMinPrice(); price < 5; price += 0.001) {
            var current = plugin.round(price).getAsDouble();
            assertTrue(previous <= current);
            previous = current;
        }
    }

}

File: mns-reactor/src/test/java/cz/fb/manaus/reactor/rounding/MatchbookRoundingPluginTest.kt
package cz.fb.manaus.reactor.rounding

import cz.fb.manaus.core.model.provider
import cz.fb.manaus.core.test.AbstractLocalTestCase
import cz.fb.manaus.spring.ManausProfiles.TEST
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.context.ActiveProfiles

@ActiveProfiles(value = ["matchbook", TEST], inheritProfiles = false)
class MatchbookRoundingPluginTest : AbstractLocalTestCase() {

    @Autowired
    private lateinit var pluginStep: RateStepRoundingPlugin


    @Test
    fun `shift by 1`() {
        assertEquals(2.02, pluginStep.shift(2.0, 1), 0.0001)
    }

    @Test
    fun `inc 1_X prices`() {
        assertEquals(1.227, pluginStep.shift(1.222, 1), 0.001)
        assertEquals(1.128, pluginStep.shift(1.125, 1), 0.001)
        assertEquals(1.131, pluginStep.shift(1.128, 1), 0.001)
        assertEquals(1.13, pluginStep.shift(1.125, 2), 0.001)
        assertEquals(2.01, pluginStep.shift(1.99, 1), 0.001)
    }

    @Test
    fun `dec 2_X prices`() {
        assertEquals(1.98, pluginStep.shift(2.0, -1), 0.0001)
        assertEquals(2.764, pluginStep.shift(2.8, -1), 0.0001)
    }

    @Test
    fun `dec 3_X prices`() {
        assertEquals(2.96, pluginStep.shift(3.0, -1), 0.0001)
    }

    @Test
    fun `inc 3_X prices`() {
        assertEquals(3.55, pluginStep.shift(3.5, 1), 0.0001)
    }

    @Test
    fun `inc 5_X prices`() {
        assertEquals(5.08, pluginStep.shift(5.0, 1), 0.0001)
    }

    @Test
    fun `rounding - big range`() {
        var previous = -1.0
        var price = provider.minPrice
        while (price < 5) {
            val current = pluginStep.round(price)
            assertTrue(previous <= current)
            previous = current
            price += 0.001
        }
    }

}